File: samples/accounts/2-withdraw-money/fsharp/Program.fs
open System
open System.IO
open Microsoft.Data.Sqlite
open Microsoft.Extensions.Configuration
open Microsoft.Extensions.Logging
open FCQRS.Actor
open FCQRS.Common
open FCQRS.FSharp
open Account

let database = Path.Combine(AppContext.BaseDirectory, "accounts.db")

// FCQRS logs through Microsoft.Extensions.Logging; this sample stays quiet.
let logging = LoggerFactory.Create(fun _ -> ())
// FCQRS reads optional settings from IConfiguration; this sample sets none.
let configuration = ConfigurationBuilder().Build()
// Events go to a SQLite file next to the program.
let connection = Fcqrs.connect DBType.Sqlite $"Data Source={database};"
let api = Fcqrs.actor configuration logging (Some connection) "accounts"

// Register the account rules; `accounts` sends commands to them.
let accounts =
    Fcqrs.aggregate api
        { Name = "Account"
          Initial = initial
          Decide = decide
          Fold = fold
          Snapshots = Default
          Passivation = PassivationPolicy.Default }

// Finish startup. This program has no sagas yet.
Fcqrs.wireSagaStarters api []

let describe event =
    match event with
    | Opened owner -> $"Opened for {owner}"
    | Deposited amount -> $"Deposited {amount}"
    | Withdrawn amount -> $"Withdrew {amount}"
    | Rejected reason -> $"Rejected: {reason}"

// docs:send
// Each account ID gets its own aggregate instance.
let alice = Fcqrs.aggregateId "alice"

// Send a command and return the account's reply.
let request command =
    accounts.Send (Fcqrs.newCid ()) alice command (fun _ -> true)

// Print a reply. Journaled tells whether FCQRS stored the event.
let show (reply: Event<AccountEvent>) =
    let stored = if reply.Journaled = Some true then "stored" else "not stored"
    let description = describe reply.EventDetails
    printfn $"{description} (version {reply.Version}, {stored})"

let send command = request command |> Async.RunSynchronously |> show

send (Open "Alice")
send (Deposit 100m)
send (Withdraw 30m)
send (Withdraw 500m)
send (Open "Alice")
// docs:end

// docs:together
// Two withdrawals of 60 arrive at the same moment.
let replies =
    [ request (Withdraw 60m); request (Withdraw 60m) ]
    |> Async.Parallel
    |> Async.RunSynchronously

// Print the stored one first.
replies
|> Array.sortBy (fun reply -> reply.Journaled <> Some true)
|> Array.iter show
// docs:end

// Applications read stored events through projections (step 4). This reads the
// journal table directly only to show what FCQRS stored.
let printJournal () =
    use connection = new SqliteConnection($"Data Source={database}")
    connection.Open()
    use query = connection.CreateCommand()
    query.CommandText <-
        "SELECT sequence_number, json_extract(CAST(message AS TEXT), '$.EventDetails') FROM journal
         WHERE persistence_id = 'Account/default-shard/alice' ORDER BY sequence_number"
    use rows = query.ExecuteReader()
    printfn "\nJournal rows for Account/default-shard/alice:"
    while rows.Read() do
        printfn "  %d  %s" (rows.GetInt64 0) (rows.GetString 1)

printJournal ()
api.Stop().Wait()

File: samples/accounts/2-withdraw-money/csharp/Program.cs
using FCQRS;
using Microsoft.Data.Sqlite;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;
using static FCQRS.Common;
using static FCQRS.CSharp;

var database = Path.Combine(AppContext.BaseDirectory, "accounts.db");

var builder = Host.CreateApplicationBuilder();
builder.Logging.ClearProviders();
// Events go to a SQLite file next to the program; register the account rules.
builder.Services.AddFcqrs($"Data Source={database};", "accounts")
    .AddAggregate<Account>();
using var host = builder.Build();
await host.StartAsync();

// docs:send
// Sends commands to accounts and returns the event each one replied with.
var accounts = host.Services
    .GetRequiredService<Handler<AccountCommand, AccountEvent>>();
// Each account ID gets its own aggregate instance.
var alice = Values.CreateAggregateId("alice");

// Send a command and return the account's reply.
Task<Event<AccountEvent>> Request(AccountCommand command) =>
    accounts(_ => true, Values.NewCID(), alice, command);

// Print a reply. Journaled tells whether FCQRS stored the event.
void Show(Event<AccountEvent> reply)
{
    var stored = reply.Journaled?.Value == true ? "stored" : "not stored";
    var description = Describe(reply.EventDetails);
    Console.WriteLine($"{description} (version {reply.Version}, {stored})");
}

async Task Send(AccountCommand command) => Show(await Request(command));

await Send(new Open("Alice"));
await Send(new Deposit(100m));
await Send(new Withdraw(30m));
await Send(new Withdraw(500m));
await Send(new Open("Alice"));
// docs:end

// docs:together
// Two withdrawals of 60 arrive at the same moment.
var replies = await Task.WhenAll(
    Request(new Withdraw(60m)), Request(new Withdraw(60m)));

// Print the stored one first.
foreach (var reply in replies.OrderBy(reply => reply.Journaled?.Value != true))
    Show(reply);
// docs:end

PrintJournal(database);
await host.StopAsync();

static string Describe(AccountEvent @event) => @event switch
{
    Opened opened => $"Opened for {opened.Owner}",
    Deposited deposited => $"Deposited {deposited.Amount}",
    Withdrawn withdrawn => $"Withdrew {withdrawn.Amount}",
    Rejected rejected => $"Rejected: {rejected.Reason}"
};

// Applications read stored events through projections (step 4). This reads the
// journal table directly only to show what FCQRS stored.
static void PrintJournal(string database)
{
    using var connection = new SqliteConnection($"Data Source={database}");
    connection.Open();
    using var query = connection.CreateCommand();
    query.CommandText =
        """
        SELECT sequence_number, json_extract(CAST(message AS TEXT), '$.EventDetails') FROM journal
        WHERE persistence_id = 'Account/default-shard/alice' ORDER BY sequence_number
        """;
    using var rows = query.ExecuteReader();
    Console.WriteLine();
    Console.WriteLine("Journal rows for Account/default-shard/alice:");
    while (rows.Read())
        Console.WriteLine($"  {rows.GetInt64(0)}  {rows.GetString(1)}");
}
